File: src/design_paradigms/divide_and_conquer/quick_sort.py
def quick_sort(arr):
    """
    Perform quick sort on an array.

    Parameters:
        arr (list): The list of elements to be sorted.

    Returns:
        list: A new sorted list.
    """
    if len(arr) <= 1:
        return arr

    # Choose the pivot (here we use the last element)
    pivot = arr[-1]
    left = [x for x in arr[:-1] if x <= pivot]  # Elements less than or equal to pivot
    right = [x for x in arr[:-1] if x > pivot]  # Elements greater than pivot

    # Recursively sort left and right partitions and combine with pivot
    return quick_sort(left) + [pivot] + quick_sort(right)

File: src-tauri/src/Algorithms/divide_and_conquer/quick_sort.rs
fn quick_sort<T: Ord>(arr: &mut [T]) {
    if arr.len() <= 1 {
        return; // Base case: already sorted
    }

    let pivot_index = partition(arr);
    let (left, right) = arr.split_at_mut(pivot_index);
    
    quick_sort(left);          // Recursively sort the left part
    quick_sort(&mut right[1..]); // Recursively sort the right part (excluding pivot)
}

fn partition<T: Ord>(arr: &mut [T]) -> usize {
    let pivot_index = arr.len() / 2; // Choose middle element as pivot
    arr.swap(pivot_index, arr.len() - 1); // Move pivot to the end
    
    let mut i = 0;
    for j in 0..arr.len() - 1 {
        if arr[j] <= arr[arr.len() - 1] { // Compare with pivot
            arr.swap(i, j);
            i += 1;
        }
    }
    
    arr.swap(i, arr.len() - 1); // Move pivot to its final position
    i
}
